Add IdpConfirmOverrideLinkAuthenticator to handle duplicate federated identity

lexcao opened the pull request:
Closes #26201.

This PR introduces a new Authenticator for broker first login flow to display a page where users can confirm overriding the existing broker link.

The page looks like this.


<img width="720" alt="image" src="https://github.com/keycloak/keycloak/assets/14901139/2e491f2d-2473-4c03-949c-da320ddedb05">

lexcao commented:
Hi @mposolda and @pedroigor 
The PR is ready for review, I will create another PR to fix the email issue before merging(see TODO for reference).

After finish this PR, I find there are some common logic to the `IdpConfirmLinkAuthenticator`.
I prefer extending the `IdpConfirmLinkAuthenticator` to remove code duplication, wdyt?

lexcao commented:
> I will create another PR to fix the email issue before merging(see TODO for reference).

Created, this PR should be merged after this issue #26418

mposolda commented on services/src/main/java/org/keycloak/authentication/authenticators/broker/IdpEmailVerificationAuthenticator.java:
```diff
@@ -115,6 +115,8 @@ private void sendVerifyEmail(KeycloakSession session, AuthenticationFlowContext
         UriInfo uriInfo = session.getContext().getUri();
         AuthenticationSessionModel authSession = context.getAuthenticationSession();
+        // TODO: split to a new PR
+        authSession.addRequiredAction(UserModel.RequiredAction.VERIFY_EMAIL);
```
Is it correct, that this should be now removed from this PR since we already have https://github.com/keycloak/keycloak/commit/142c1413 ?

mposolda commented on themes/src/main/resources/theme/base/login/messages/messages_en.properties:
```diff
@@ -269,11 +269,14 @@ federatedIdentityUnavailableMessage=User {0} authenticated with identity provide
 confirmLinkIdpTitle=Account already exists
+confirmOverrideIdpTitle=Broker link already exists
 federatedIdentityConfirmLinkMessage=User with {0} {1} already exists. How do you want to continue?
+federatedIdentityConfirmOverrideMessage=It seems that the account you are trying to link has already been linked to another account using the identity provider {0}. Can you confirm if you want to replace the existing link with the new account?
```
I am not 100% sure if this message is a bit confusing? It is going to be shown to the end user, so perhaps we might come with something more clear, so the user is more clearly aware of what is going on? I can think about something like:
```
You are trying to link your account {0} with the {1} account {2}. But your account is already linked with different {3} account {4}. Can you confirm if you want to replace the existing link with the new account?
```
{0} - Keycloak username
{1} - idp alias
{2} - new federatedIdentity username
{3} - idp alias (same like {1})
{4} - old federatedIdentity username (something like `federatedIdentity.getUsername()` from existing link)

Is it possible? If it is a headache to display existing link, we can possibly skip {4} in the message above (as I am not sure if it requires to add new freemarker bean to display the info about existing link details, but I am not sure if it worth it...).

@pedroigor WDYT?

pedroigor replied:
Yeah, your message gives more information and makes clearer what is going on.

mposolda reviewed (commented):
@lexcao Thanks! Added 2 review comments inline. I would like to review also from @pedroigor for this PR.

Besides that, I think we need a documentation for this? Likely update some existing section under https://www.keycloak.org/docs/latest/server_admin/index.html#_identity_broker_first_login or add new section here?

pedroigor reviewed (commented):
@lexcao @mposolda LGTM.

pedroigor reviewed (commented):
@lexcao @mposolda Does it make sense to fire a specific event and make it easier to monitor attempts to override a federation link? It can help tracking this form of user activity that can potentially cause issues with their account ...

lexcao commented:
@mposolda 
For the documentation, I will take care of it.

@pedroigor 
The event is a good idea, I vote for that.
There could be an event named `FEDERATED_IDENTITY_OVERRIDE_LINK`.

lexcao commented:
Hi @mposolda 
By the way, I created a bunch new files, should I add the license at the beginning of the file manually?

mposolda commented:
@pedroigor @lexcao +1 for the event.

Yes, adding copyright headers can be good (I don't think it is strictly required, but can be nice). It should be possible to automatically create some template in your IDE, so you don't need to add it manually. For example Intellij Idea has it in: `File` -> `Settings` -> `Copyright` -> `Copyright profiles` . Then creating some profile with the text like this:
```
Copyright 2024 Red Hat, Inc. and/or its affiliates
 and other contributors as indicated by the @author tags.
 
 Licensed under the Apache License, Version 2.0 (the "License");
 you may not use this file except in compliance with the License.
 You may obtain a copy of the License at
 
 http://www.apache.org/licenses/LICENSE-2.0
 
 Unless required by applicable law or agreed to in writing, software
 distributed under the License is distributed on an "AS IS" BASIS,
 WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 
 See the License for the specific language governing permissions and
 limitations under the License.
```

lexcao commented:
@mposolda  
Thanks for the suggestion on the copyright headers.

I have added the event and updated the error message, updated the screenshot as well.

The documentation is also added, but I don't know what content needed to be include.
Please give the suggestions on the documentation.

vincentpalita commented:
Hi! Any chance this gets accepted and shipped into v25?

mposolda commented on docs/documentation/server_admin/topics/identity-broker/first-login-flow.adoc:
```diff
@@ -115,3 +115,10 @@ You could set the also set `Sync Mode` to `force` if you want to update the user
+
+[[override_existing_broker_link]]
+==== Override existing broker link
+When an another account needs to be linked to the same Keycloak account within the same identity provider, you can configure the following authenticator.
```
```suggestion
When an another account needs to be linked to the same {project_name} account within the same identity provider, you can configure the following authenticator.
```

mposolda commented on docs/documentation/server_admin/topics/identity-broker/first-login-flow.adoc:
```diff
@@ -115,3 +115,10 @@ You could set the also set `Sync Mode` to `force` if you want to update the user
+When an another account needs to be linked to the same Keycloak account within the same identity provider, you can configure the following authenticator.
+
+Confirm Override Existing Link::
+This authenticator will detect the existing broker link for the user and display a confirmation page to confirm overriding the existing broker link. Set the authenticator requirement to REQUIRED.
```
Is it possible to add few additional sentences to make it a bit more clear how to use this authenticator and about the usage of this? I can think about adding something like this.

```
The typical use of this authenticator is for instance the scenario like this one: 
* You have for instance existing {project_name} user `john` with email `john@gmail.com`, who is already linked with identity provider `google` with `google` username being `john@gmail.com` .
* Then for instance {project_name} user `john` updates his email in {project_name} to `john-new@gmail.com`
* Then during login to {project_name}, the user authenticated to identity provider `google` with his new username like `john-new@gmail.com`, which is not linked to any {project_name} account yet (as {project_name} account `john` is still linked with `google` user `john@gmail.com`) and hence first-broker-login flow is triggered.
* During first-broker-login, the {project_name} user `john` is authenticated somehow (either by default first-broker-login re-authentication or for instance by authenticator like `Detect existing broker user`)
* Now with this authenticator in the authentication flow, it is possible to override existing IDP link to `google` identity provider of {project_name} user `john` with the new `google` link to `google` user `john-new@gmail.com` after user `john` confirms this.

When creating authentication flow with this authenticator, please make sure to add this authenticator once other authenticators already established {project_name} user by other means (either by re-authentication or after `Detect existing broker user` as mentioned above.
```

mposolda reviewed (commented):
@lexcao Sorry for late review.

I hope we can have this PR ready for Keycloak 25. Added a comment regarding documentation. Also I think rebase is needed due some conflict, but hopefully would be easy to solve.

lexcao committed:
Add IdpConfirmOverrideLinkAuthenticator to handle duplicate federated identity (#26201)

Closes #26201.

This PR introduces a new Authenticator for broker first login flow to display a page where users can confirm overriding the existing broker link.

Signed-off-by: Lex Cao <lexcao@foxmail.com>

lexcao commented:
Hi @mposolda 
Thanks for pushing this up and content suggestion, you are awesome!
I have updated the code.

Let's wait the CI to pass.

mposolda reviewed (dismissed):
@lexcao Thanks!

I am adding @andymunro to review the documentation part. If he is OK, I hope to finally merge this PR.

andymunro commented on docs/documentation/server_admin/topics/identity-broker/first-login-flow.adoc:
```diff
@@ -115,3 +115,19 @@ You could set the also set `Sync Mode` to `force` if you want to update the user
+Confirm Override Existing Link::
+This authenticator will detect the existing broker link for the user and display a confirmation page to confirm overriding the existing broker link. Set the authenticator requirement to REQUIRED.
+
+The typical use of this authenticator is for instance the scenario like this one:
```
```suggestion
A typical use of this authenticator is a scenario such as the following:
```

andymunro commented on docs/documentation/server_admin/topics/identity-broker/first-login-flow.adoc:
```diff
@@ -115,3 +115,19 @@ You could set the also set `Sync Mode` to `force` if you want to update the user
+This authenticator will detect the existing broker link for the user and display a confirmation page to confirm overriding the existing broker link. Set the authenticator requirement to REQUIRED.
+
+The typical use of this authenticator is for instance the scenario like this one:
+* You have for instance existing {project_name} user `john` with email `john@gmail.com`, who is already linked with identity provider `google` with `google` username being `john@gmail.com` .
```
```suggestion
* For example, consider a {project_name} user `john` with the email `john@gmail.com`. That user is linked to the identity provider `google` with the `google` username `john@gmail.com` .
```

andymunro commented on docs/documentation/server_admin/topics/identity-broker/first-login-flow.adoc:
```diff
@@ -115,3 +115,19 @@ You could set the also set `Sync Mode` to `force` if you want to update the user
+The typical use of this authenticator is for instance the scenario like this one:
+* You have for instance existing {project_name} user `john` with email `john@gmail.com`, who is already linked with identity provider `google` with `google` username being `john@gmail.com` .
+* Then for instance {project_name} user `john` updates his email in {project_name} to `john-new@gmail.com`
+* Then during login to {project_name}, the user authenticated to identity provider `google` with his new username like `john-new@gmail.com`, which is not linked to any {project_name} account yet (as {project_name} account `john` is still linked with `google` user `john@gmail.com`) and hence first-broker-login flow is triggered.
```
```suggestion
* Then during login to {project_name}, the user authenticated to the identity provider `google` with a new username such as `john-new@gmail.com`, which is not linked to any {project_name} account yet (as {project_name} account `john` is still linked with the `google` user `john@gmail.com`) and hence the first-broker-login flow is triggered.
```

andymunro commented on docs/documentation/server_admin/topics/identity-broker/first-login-flow.adoc:
```diff
@@ -115,3 +115,19 @@ You could set the also set `Sync Mode` to `force` if you want to update the user
+* Then for instance {project_name} user `john` updates his email in {project_name} to `john-new@gmail.com`
+* Then during login to {project_name}, the user authenticated to identity provider `google` with his new username like `john-new@gmail.com`, which is not linked to any {project_name} account yet (as {project_name} account `john` is still linked with `google` user `john@gmail.com`) and hence first-broker-login flow is triggered.
+* During first-broker-login, the {project_name} user `john` is authenticated somehow (either by default first-broker-login re-authentication or for instance by authenticator like `Detect existing broker user`)
+* Now with this authenticator in the authentication flow, it is possible to override existing IDP link to `google` identity provider of {project_name} user `john` with the new `google` link to `google` user `john-new@gmail.com` after user `john` confirms this.
```
```suggestion
* Now with this authenticator in the authentication flow, it is possible to override the IDP link to the `google` identity provider of {project_name} user `john` with the new `google` link to `google` user `john-new@gmail.com` after user `john` confirms this.
```

andymunro commented on docs/documentation/server_admin/topics/identity-broker/first-login-flow.adoc:
```diff
@@ -115,3 +115,19 @@ You could set the also set `Sync Mode` to `force` if you want to update the user
+* During first-broker-login, the {project_name} user `john` is authenticated somehow (either by default first-broker-login re-authentication or for instance by authenticator like `Detect existing broker user`)
+* Now with this authenticator in the authentication flow, it is possible to override existing IDP link to `google` identity provider of {project_name} user `john` with the new `google` link to `google` user `john-new@gmail.com` after user `john` confirms this.
+
+When creating authentication flow with this authenticator, please make sure to add this authenticator once other authenticators already established {project_name} user by other means (either by re-authentication or after `Detect existing broker user` as mentioned above.
```
```suggestion
When creating authentication flows with this authenticator, make sure to add this authenticator once other authenticators that are already established the {project_name} user by other means (either by re-authentication or after `Detect existing broker user` as mentioned above.
```

andymunro reviewed (changes requested):
A few suggestions @lexcao

lexcao committed:
Apply suggestions from code review

Co-authored-by: andymunro <48995441+andymunro@users.noreply.github.com>
Signed-off-by: Lex Cao <lexcao@foxmail.com>

lexcao commented:
Hi @andymunro 
Thanks for your suggestions! All done.

mposolda reviewed (approved):
@lexcao @andymunro @jonkoops Thanks for the reviews and updates to this PR!
